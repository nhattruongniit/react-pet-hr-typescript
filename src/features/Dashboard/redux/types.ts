export type ITodo = {
  id: number;
  title: string;
};

export type IDashboardState = {
  todos: ITodo[];
  error?: boolean;
};

export type IDashboardAction = {
  type: string;
  payload: IDashboardState;
};

export enum IDashboardTypes {
  FETCH_TODO_REQUEST = 'APP/FETCH_TODO_REQUEST',
  FETCH_TODO_SUCCESS = 'APP/FETCH_TODO_SUCCESS',
  FETCH_TODO_ERROR = 'APP/FETCH_TODO_ERROR',
}
