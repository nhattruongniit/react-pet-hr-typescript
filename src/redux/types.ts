export type IAppState = {
  isSidebar: boolean;
  mode: string;
  isLoading: false
};

export type IAppAction = {
  type: string;
  payload: IAppState;
};

export enum AppActionTypes {
  // toggle side bar
  SET_SIDEBAR = 'APP/SET_SIDEBAR',
  // dark mode
  SET_DARK_MODE = 'APP/SET_DARK_MODE',
  SET_LOADING = 'APP/SET_LOADING'
}
