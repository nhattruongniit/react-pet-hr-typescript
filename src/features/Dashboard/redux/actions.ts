import { IDashboardTypes, IDashboardState } from './types';

export const fetchTodoRequest = () => ({
  type: IDashboardTypes.FETCH_TODO_REQUEST,
});

export const fetchTodoSuccess = ({ todos }: IDashboardState) => {
  return {
    type: IDashboardTypes.FETCH_TODO_SUCCESS,
    payload: { todos },
  };
};

export const fetchTodoError = () => ({
  type: IDashboardTypes.FETCH_TODO_ERROR,
});
