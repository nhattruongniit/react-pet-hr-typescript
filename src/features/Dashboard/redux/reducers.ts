// types
import { IDashboardTypes, IDashboardState, IDashboardAction } from './types';

export const initialState: IDashboardState = {
  todos: [],
  error: false,
};

const reducers = (state = initialState, { type, payload }: IDashboardAction) => {
  switch (type) {
    case IDashboardTypes.FETCH_TODO_REQUEST:
      return {
        ...state,
      };
    case IDashboardTypes.FETCH_TODO_SUCCESS:
      return {
        ...state,
        todos: payload.todos,
      };
    case IDashboardTypes.FETCH_TODO_ERROR:
      return {
        ...state,
        error: true,
      };
    default:
      return state;
  }
};

export default reducers;
