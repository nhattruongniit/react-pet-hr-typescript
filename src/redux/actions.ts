import { AppActionTypes } from './types';

// toggle sidebar
export const setSidebar = (isSidebar: boolean) => ({
  type: AppActionTypes.SET_SIDEBAR,
  payload: { isSidebar },
});

// change dark mode
export const setDarkMode = (mode: string) => ({
  type: AppActionTypes.SET_DARK_MODE,
  payload: { mode },
});
