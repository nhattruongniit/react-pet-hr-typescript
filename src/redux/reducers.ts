// types
import { AppActionTypes, IAppState, IAppAction } from './types';

const initialState: IAppState = {
  isSidebar: false,
  mode: 'light',
};

const reducers = (state = initialState, { type, payload }: IAppAction) => {
  switch (type) {
    case AppActionTypes.SET_SIDEBAR:
      return {
        ...state,
        isSidebar: payload.isSidebar,
      };
    case AppActionTypes.SET_DARK_MODE:
      return {
        ...state,
        mode: payload.mode,
      };
    default:
      return state;
  }
};

export default reducers;
