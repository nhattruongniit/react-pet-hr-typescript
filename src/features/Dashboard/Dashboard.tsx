import React, { useEffect } from 'react';
import { connect, ConnectedProps } from 'react-redux';
import { useTranslation } from 'react-i18next';

// material core
import { makeStyles } from '@material-ui/core/styles';

// components
import TypographyBase from 'components/atoms/TypographyBase';

// types
import IState from 'IState';
import { ITodo } from './redux/types';

// actions
import { fetchTodoRequest } from './redux/actions';

// styles
const useStyles = makeStyles(() => ({
  root: {
    paddingBottom: 20,
  },
}));

// connect redux
const mapStateToProps = (state: IState) => {
  const {
    dashboard: { todos },
  } = state;

  return {
    todos,
  };
};

const mapDispatchToProps = {
  fetchTodoRequest,
};

const connector = connect(mapStateToProps, mapDispatchToProps);

type PropsFromRedux = ConnectedProps<typeof connector>;

const Dashboard = ({ todos, fetchTodoRequest }: PropsFromRedux) => {
  const classes = useStyles();
  const { t: translate } = useTranslation();

  useEffect(() => {
    fetchTodoRequest();
  }, [fetchTodoRequest]);

  return (
    <div className={classes.root}>
      <TypographyBase variant="h1" color="textSecondary" text={translate('dashboard')} /> <br />
      <TypographyBase variant="h5" color="textSecondary" text="Example fetch data with saga" /> <br />
      <TypographyBase variant="h4" color="textSecondary" text="Todo" />
      <br />
      {todos.map((todo: ITodo) => (
        <div key={todo.id}>
          <TypographyBase variant="h5" color="textSecondary" text={`Id: ${todo.id.toString()}`} />
          <TypographyBase variant="h5" color="textSecondary" text={`Title: ${todo.title}`} />
        </div>
      ))}
      <br />
    </div>
  );
};

export default connector(Dashboard);
