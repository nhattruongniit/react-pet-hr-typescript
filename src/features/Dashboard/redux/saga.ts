import { call, put, takeEvery } from 'redux-saga/effects';

// types
import { IDashboardTypes } from './types';

// api
import * as api from '../api';
import { fetchTodoSuccess, fetchTodoError } from './actions';

export const fetchTodoSaga = function* () {
  try {
    const todos = yield call(api.fetchTodo);
    yield put(fetchTodoSuccess({ todos }));
  } catch (err) {
    yield put(fetchTodoError());
  }
};

const dashboardSaga = function* () {
  yield takeEvery(IDashboardTypes.FETCH_TODO_REQUEST, fetchTodoSaga);
};

export default dashboardSaga;
