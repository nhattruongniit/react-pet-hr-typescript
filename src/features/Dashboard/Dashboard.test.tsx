import React from 'react';
import renderer from 'react-test-renderer';
import configureStore from 'redux-mock-store';
import { Provider } from 'react-redux';

// components
import Dashboard from './Dashboard';

// actions
import { fetchTodoRequest } from './redux/actions';
import { IDashboardTypes } from './redux/types';

const mockStore = configureStore([]);

const initialState = {
  todos: [
    {
      id: 1,
      title: 'abc',
    },
  ],
  error: false,
};

describe('DASHBOARD COMPONENT', () => {
  let store: any = null;
  let wrapper: any = null;

  beforeEach(() => {
    store = mockStore({
      dashboard: initialState,
    });

    wrapper = renderer.create(
      <Provider store={store}>
        <Dashboard />
      </Provider>,
    );
  });

  it('SHOULD SNAPSHOT RENDER WITH CONNECTED REDUX STORE', () => {
    expect(wrapper.toJSON()).toMatchSnapshot();
  });

  it('SHOULD DISPATCH ACTION REQUEST FETCHING TODO AFTER RENDER COMPONENT', () => {
    expect(fetchTodoRequest()).toEqual({ type: IDashboardTypes.FETCH_TODO_REQUEST });
  });
});
