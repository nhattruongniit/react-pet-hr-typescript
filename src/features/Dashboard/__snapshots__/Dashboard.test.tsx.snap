// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`DASHBOARD COMPONENT SHOULD SNAPSHOT RENDER WITH CONNECTED REDUX STORE 1`] = `
<div
  className="makeStyles-root-1"
>
  <div
    className="MuiTypography-root MuiTypography-h1 MuiTypography-colorTextSecondary"
  >
    dashboard
  </div>
   
  <br />
  <div
    className="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary"
  >
    Example fetch data with saga
  </div>
   
  <br />
  <div
    className="MuiTypography-root MuiTypography-h4 MuiTypography-colorTextSecondary"
  >
    Todo
  </div>
  <br />
  <div>
    <div
      className="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary"
    >
      Id: 1
    </div>
    <div
      className="MuiTypography-root MuiTypography-h5 MuiTypography-colorTextSecondary"
    >
      Title: abc
    </div>
  </div>
  <br />
</div>
`;
